File: web/src/types/featureFlags.ts
import { Routes } from '../generated/contracts';

/**
 * Phase 8 (Cross-Platform SSOT plan) remote kill-switch payload -- the wire
 * shape of `feature_flags/config`. Mirrors KMP's `FeatureFlags` domain model
 * (`shared/.../domain/model/FeatureFlags.kt`) field-for-field.
 */
export interface FeatureFlags {
  minimumSupportedAppVersion: string;
  flags: Record<string, boolean>;
}

/**
 * Fail-open fallback: the version baked into this build at build time
 * (`contracts/routes.yaml`), used whenever the live Firestore doc is
 * unreachable or missing. Never blocks on its own.
 */
export const SAFE_DEFAULT_FEATURE_FLAGS: FeatureFlags = {
  minimumSupportedAppVersion: Routes.MINIMUM_SUPPORTED_APP_VERSION,
  flags: {}
};

/**
 * Compares dotted major.minor.patch version strings (numeric per-segment
 * compare, not lexicographic). Missing or non-numeric segments are treated
 * as 0. Mirrors KMP's `isAppVersionBelowMinimum` exactly -- see that
 * function's doc comment for why malformed input can't throw here either.
 */
export function isAppVersionBelowMinimum(current: string, minimum: string): boolean {
  const currentParts = current.split('.').map((s) => parseInt(s, 10) || 0);
  const minParts = minimum.split('.').map((s) => parseInt(s, 10) || 0);
  const length = Math.max(currentParts.length, minParts.length);
  for (let i = 0; i < length; i++) {
    const c = currentParts[i] ?? 0;
    const m = minParts[i] ?? 0;
    if (c !== m) return c < m;
  }
  return false;
}

File: shared/src/commonMain/kotlin/com/ssbmax/shared/domain/model/FeatureFlags.kt
package com.ssbmax.shared.domain.model

import com.ssbmax.shared.contracts.SsbContracts

/**
 * Phase 8 (Cross-Platform SSOT plan) remote kill-switch payload -- read from
 * `feature_flags/config` at launch by every platform, cached with
 * [SAFE_DEFAULT] so a Firestore outage never blocks a user (fail-open, not
 * fail-closed: an indie developer cannot remotely *lock everyone out* by
 * accident, only remotely *raise the floor*).
 */
data class FeatureFlags(
    val minimumSupportedAppVersion: String = SsbContracts.Routes.MINIMUM_SUPPORTED_APP_VERSION,
    val flags: Map<String, Boolean> = emptyMap()
) {
    fun isEnabled(name: String, default: Boolean = false): Boolean = flags[name] ?: default

    companion object {
        /**
         * Fail-open fallback: the version baked into this build at compile
         * time (contracts/routes.yaml), used whenever the live Firestore doc
         * is unreachable or missing. Never blocks on its own.
         */
        val SAFE_DEFAULT = FeatureFlags()
    }
}

/**
 * Compares dotted major.minor.patch version strings (e.g. "1.2.10" > "1.2.9",
 * ordinary numeric compare per segment, not lexicographic). Missing or
 * non-numeric segments are treated as 0, so a malformed version string can
 * never throw -- it just compares as if that segment were absent.
 */
fun isAppVersionBelowMinimum(current: String, minimum: String): Boolean {
    val currentParts = current.split(".").map { it.toIntOrNull() ?: 0 }
    val minParts = minimum.split(".").map { it.toIntOrNull() ?: 0 }
    for (i in 0 until maxOf(currentParts.size, minParts.size)) {
        val c = currentParts.getOrElse(i) { 0 }
        val m = minParts.getOrElse(i) { 0 }
        if (c != m) return c < m
    }
    return false
}
